### PR TITLE
[mlir-hlo] Added RegionBranchOpInterfaces to lmhlo operations.

### PR DESCRIPTION
**Pull request by dfki-jugr**
Added RegionBranchOpInterfaces to lmhlo operations that use regions.
This is needed, since the bufferization features in MLIR have to reason about the control flow within these operations.

### CODE REVIEW

**Review by timshen91 — APPROVED** [review]

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/lhlo_ops.cc [hunk 1]
```diff
@@ -226,6 +257,24 @@ static LogicalResult Verify(CustomCallOp op) {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// MapOp
+//===----------------------------------------------------------------------===//
+
+void MapOp::getSuccessorRegions(Optional<unsigned> index,
```
Comment (sanjoy) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/lhlo_ops.cc:264]: I don't think the region attached to Map is the successor of the operation (as defined [here](https://github.com/llvm/llvm-project/blob/92880ab7a2b2145f0605f367cd6d53d6892903c3/mlir/include/mlir/Interfaces/ControlFlowInterfaces.td#L130)).  Rather, is part of the operation definition, like convolution dim numbers are for a convolution.

I'm new to this so please push back if I'm not making sense.

**Review by sanjoy — CHANGES_REQUESTED** [review]

**Review by timshen91 — CHANGES_REQUESTED** [review]
Now that I take another look, yes, I agree with Sanjoy's comment. For the purpose of bufferization, RegionBranchOpInterfaces should be added to only lmhlo.while and lmhlo.case. Other ops are with regions that don't really allocate.

**Comment by dfki-mako** [comment]
@timshen91 Several analyses and runs within MLIR were designed with the `RegionBranchOpInterface` in mind. One of the underlying ideas was that any operation that has at least one region (excluding operations that do not return results) must implement this interface. If this invariant does not hold in general, some MLIR control-flow-based analyses cannot be successfully applied (including `BufferDeallocation`). This raises the question of whether this PR should also add "proper" implementations of `RegionBranchOpInterface` to the "remaining" operations....

On the one hand, one can argue that the implementation of `RegionBranchOpInterface` is generally not required for operations that have only a single region (as outlined above). On the other hand, it can be argued that a region attached to, say, the `Map` operation is called successively until each element is transformed from a high-level point of view in theory. This would form an implicit loop "hidden" within the operation. Alternatively, you can claim that this does not form a loop in general and should not be "exposed to other transformations". What do you think about this?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/lhlo_ops.cc (hunk 1 above)
Comment (sherhut) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/lhlo_ops.cc:264]: That is indeed an interesting question. I think the original modelling of these interfaces had classical conditionals and loops in mind and that also shows in the doc comments.

However, I still think this is modeled correctly here. The body of the map operation has to execute before the map operation can complete. In that sense, the body region is the successor of the outer map operation. This does not cover the fact that it potentially runs multiple times in parallel etc. and that it has an implicit selection of elements from the operands.

An alternative would be to declare the body region as `IsolatedFromAbove` to give it more function-like semantics. That would in essence turn the map body into its own scope for allocation. I don't think that is supported, though. Currently, none of this matters as the map function is on scalars and hence typically does not allocate.

The same is true for the other ops, except for while, case, cond and fusion op, where the branch op interface is the correct choice.

**Review by sherhut — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/lhlo_ops.cc (hunk 1 above)
Comment (joker-eph) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/lhlo_ops.cc:264]: I don't quite get why the region isn't the successor of the operation? What is the predecessor of the region here then?

**Comment by timshen91** [comment]
> This raises the question of whether this PR should also add "proper" implementations of `RegionBranchOpInterface` to the "remaining" operations....

I guess it depends on what exactly `RegionBranchOpInterface` is used for. As @sherhut mentioned, it's modeled after classical conditionals and sequential loops, not parallel loops (e.g. Map) or associative folding (e.g. Reduce). I'm leaning towards delaying it until we have a concrete use case, but maybe you already have one?

**Comment by sherhut** [comment]
> > This raises the question of whether this PR should also add "proper" implementations of `RegionBranchOpInterface` to the "remaining" operations....
> 
> I guess it depends on what exactly `RegionBranchOpInterface` is used for. As @sherhut mentioned, it's modeled after classical conditionals and sequential loops, not parallel loops (e.g. Map) or associative folding (e.g. Reduce). I'm leaning towards delaying it until we have a concrete use case, but maybe you already have one?

Ultimately all operations with a region need to implement this interface to model control flow. We can leave it out of this PR for now to unblock the work and make progress. But even then, if these operations appear in a program, we cannot guarantee correctness of the analysis because the regions have undefined control flow.

**Comment by dfki-mako** [comment]
Although these interfaces were originally developed to model classical control flow, we share @sherhut's point that each operation with a nested region must implement this interface. As @sherhut also mentioned, we cannot guarantee the correctness of various control flow-based analyses that could be applied to these operations.

What we could do is omit these changes for now, as suggested. However, we should think about implementing this interface in a follow-up PR to avoid possible problems with control-flow-based transformations in the future.

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/lhlo_ops.cc (hunk 1 above)
Comment (dfki-mako) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/lhlo_ops.cc:264]: @joker-eph I have the same question regarding the predecessor of this region.

**Comment by dfki-mako** [comment]
After a short offline discussion with @sherhut about different possibilities how to implement the `RegionBranchOpInterface`, we think that we first include interface implementations for the "basic" control-flow-dependent operations in the scope of this PR. Further extensions can be discussed in a separate issue 🤓

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/lhlo_ops.cc [hunk 2]
```diff
@@ -310,5 +358,17 @@ void FusionOp::build(OpBuilder& builder, OperationState& result,
...
 }
 
+void FusionOp::getSuccessorRegions(Optional<unsigned> index,
+                                ArrayRef<Attribute> operands,
+                                SmallVectorImpl<RegionSuccessor> &regions) {
+  // If the predecessor is the fusion region, jump back to the parent op.
+  if (index.hasValue()) {
+    assert(index.getValue() == 0 && "expected fusion region");
+    regions.push_back(RegionSuccessor());
+  } else
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/lhlo_ops.cc:368]: nit: I'd use `{` around the block here due to the comment.

**Review by sherhut — COMMENTED** [review]
Could you fix the

**Review by sherhut — APPROVED** [review]
Thanks

File: tensorflow/compiler/mlir/hlo/BUILD [hunk 3]
```diff
@@ -445,7 +448,9 @@ cc_library(
         "@llvm-project//llvm:Support",
         "@llvm-project//mlir:Analysis",
         "@llvm-project//mlir:CopyOpInterface",
+        "@llvm-project//mlir:ControlFlowInterfaces",
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/BUILD:451]: This BUILD file is ill-formatted. Move one line up. Also below.

**Review by rthadur — APPROVED** [review]

**Pull request merged**